### ISSUE TITLE
New Rule Proposal: new-native-constructor

### ISSUE THREAD

**Issue by ilyakamens** [eslint/eslint#11591]
**Please describe what the rule should do:**

If enabled, the new rule should require the `new` keyword when making use of native constructors, preventing using them as factory functions.

Practically speaking, there's no functional difference between the two. Technically, however, without `new`, we're making use of a factory function; with `new`, we're making use of the object constructor. I think the former exists more for backwards compatibility than anything else.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[x] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```
// Good
throw new Error('foo');

// Bad
throw Error('bar');

// Good
const x = new Array(1, 2, 3);

// Bad
const y = Array(1, 2, 3); 
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Seems to adhere to the core rule guidelines: https://eslint.org/docs/developer-guide/contributing/new-rules

More specifically:
1. Enforces consistency in the codebase
2. Prevents making use of rather obtuse/surprising (albeit technically standard) behavior
3. Doesn't seem like this is something specific to me, my codebase, or my organization

(Note: an alternate name could be `no-native-constructors-as-functions`)

**Are you willing to submit a pull request to implement this rule?**

Surprised me here! Sure, although I imagine it'd take me a while.

**Comment by platinumazure** [comment]
Hi @ilyakamens, thanks for the issue.

You can configure this sort of thing using [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax):

```js
{
    "no-restricted-syntax": ["error", "NewExpression[callee.name='Array']"]
}
```

Also, we do have a rule for `Object` specifically: [`no-new-object`](https://eslint.org/docs/rules/no-new-object).

**Comment by ilyakamens** [comment]
Thanks for the prompt reply @platinumazure. I didn't know about `no-restricted-syntax`; that's pretty cool.

Perhaps I'm misunderstanding, but it doesn't look like either of those rules address what I'm proposing. In fact, the first one does the opposite of what I want, since it forbids `new Array()` but allows `Array()`. I scanned the list of [AST node types](https://github.com/eslint/espree/blob/master/lib/ast-node-types.js), but it's unclear to me if any of those would work.

[Here's an example](https://github.com/nodejs/node/pull/3714/files) of the people at Node building a rule for a specific case (i.e., only `throw new Error()`) of what I've described.

**Comment by platinumazure** [comment]
@ilyakamens Oops, you're right, I made a mistake.

You would want to use CallExpression instead of NewExpression:

```js
{
    "no-restricted-syntax": ["error", "CallExpression[callee.name='Array']"]
}
```

Hope this helps!

**Comment by ilyakamens** [comment]
Beautiful. This is exactly what I was looking for. Sorry for the GH issue noise and thanks again!
